### ISSUE TITLE
After installing several toolchains and switching default one zsh completion broken

### ISSUE THREAD

**Issue by AlexCzar** [rust-lang/cargo#11437]
### Problem

I've installed several toolchains of rust, stable, 1.63.0, 1.65.0 and switched default between them several times.
Now zsh completion is completely broken for cargo.
What I tried:
* uninstall all toolchains but one
* uninstall all toolchains, install one
Neither of these helped, completion still broken.

I've found two similar issues, both of them are closed without a solution.

### Steps

1. Install several toolchains
2. switch default between them
3. type `cargo <tab>`
4. Observe:

```
; cargo error: couldn't read +stable-x86_64-unknown-linux-gnu: No such file or directory (os error 2)

error: aborting due to previous error

           
_cargo:source:2: no such file or directory: /share/zsh/site-functions/_cargo
error: couldn't read +stable-x86_64-unknown-linux-gnu: No such file or directory (os error 2)

error: aborting due to previous error

_cargo:source:2: no such file or directory: /share/zsh/site-functions/_cargo
error: couldn't read +stable-x86_64-unknown-linux-gnu: No such file or directory (os error 2)

error: aborting due to previous error

_cargo:source:2: no such file or directory: /share/zsh/site-functions/_cargo
```

### Possible Solution(s)

Completely remove everything and do a clean install (I haven't tried this yet).

### Version

```text
; cargo version --verbose
cargo 1.65.0 (4bc8f24d3 2022-10-20)
release: 1.65.0
commit-hash: 4bc8f24d3e899462e43621aab981f6383a370365
commit-date: 2022-10-20
host: x86_64-unknown-linux-gnu
libgit2: 1.5.0 (sys:0.15.0 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:OpenSSL/1.1.1q)
os: openSUSE 20221126 [64-bit]
```

**Comment by weihanglo** [comment]
Sorry about you hitting this issue. From my understanding, Cargo doesn't provide completion for shell on its own. `rustup completions` is the command doing that, but users need to configure it manually IIRC.

Could you provide the way you install Rust and toolchains, as well as how you configure shell completions? It also helps if you are able to share a repository on GitHub or via a zip file of a minimal reproducible example.

**Comment by AlexCzar** [comment]
I'm not sure how to provide MCVE here and I've found an easy enough work-around:
```bash
; rm -f $ZSH_CACHE_DIR/completions/_cargo
; ln -s $HOME/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share/zsh/site-functions/_cargo $ZSH_CACHE_DIR/completions/
```

I'm not sure how fragile this is or why it was needed, initially it just worked.
It also has a downside that when switching toolchains, to get toolchain-specific completions one needs to do this manually for the toolchain in question.
That said I'm not sure this wasn't the case to begin with.

**Comment by AlexCzar** [comment]
## TL;DR:
I've found the culprint, it has nothing to do with either cargo or rustup.

## The full story
Issue was that I had a zsh plugin for rust installed at some point which creates a clever dynamic completion setup for rustup and cargo based on the `rustup default` invocation.
I have disabled the plugin at some point while I wasn't using Rust, but the completions configuration is not removed when the plugin is disabled, so it remained there.
When I started using Rust again, since completions worked I didn't remember about the plugin and thought rustup configured the completions on install.
At some point the completion setup by the zsh plugin broke and that created the issue I misguidedly reported here.
Here's what the plugin did:
```zsh
if ! (( $+commands[rustup] && $+commands[cargo] )); then
  return
fi

# If the completion file doesn't exist yet, we need to autoload it and
# bind it to `cargo`. Otherwise, compinit will have already done that
if [[ ! -f "$ZSH_CACHE_DIR/completions/_cargo" ]]; then
  autoload -Uz _cargo
  typeset -g -A _comps
  _comps[cargo]=_cargo
fi

# If the completion file doesn't exist yet, we need to autoload it and
# bind it to `rustup`. Otherwise, compinit will have already done that
if [[ ! -f "$ZSH_CACHE_DIR/completions/_rustup" ]]; then
  autoload -Uz _rustup
  typeset -g -A _comps
  _comps[rustup]=_rustup
fi

# Generate completion files in the background
rustup completions zsh >| "$ZSH_CACHE_DIR/completions/_rustup" &|
cat >| "$ZSH_CACHE_DIR/completions/_cargo" <<'EOF'
#compdef cargo
source "$(rustc +${${(z)$(rustup default)}[1]} --print sysroot)"/share/zsh/site-functions/_cargo
EOF
```

**Comment by weihanglo** [comment]
Glad you figured it out! The write-up is really useful for someone bumping into the same issue in the future. Thank you for providing the information!
